lib/src/modules/status.cc

```diff
@@ -134,63 +143,102 @@ ActionOutcome Status::callAction(const ActionRequest& request) {
 
-    auto metadata_file = (results_dir_path / "metadata").string();
+    bool running_by_pid { false };
+    bool not_running_by_pid { false };
+    auto pid_file = (results_dir_path / "pid").string();
+    std::string pid_txt {};
+    int pid {};
+
+    // Read the PID file before the metadata, to avoid a race in the
+    // non-blocking task thread, due to the elapsed time between the
+    // end of the job and the metadata write
+    if (!fs::exists(pid_file)) {
+        LOG_DEBUG("PID file '%1%' does not exist", pid_file);
+    } else if (!lth_file::read(pid_file, pid_txt)) {
+        LOG_ERROR("Failed to read PID file '%1%'", pid_file);
+    } else if (pid_txt.empty()) {
+        LOG_ERROR("PID file '%1%' is empty", pid_file);
+    } else {
+        try {
+            pid = std::stoi(pid_txt);
+            // NOTE(ale): processExists() does not throw
+            if (Util::processExists(pid)) {
+                // NOTE(ale): will set status to 'running' iff
+                // metadata is valid and metadata.completed == false
+                running_by_pid = true;
+            } else {
+                not_running_by_pid = true;
+                bool cached { false };
+                {
+                    ResultsMutex::LockGuard a_l {
+                        ResultsMutex::Instance().access_mtx };  // RAII
+                    cached = ResultsMutex::Instance().exists(t_id);
+                }
+                if (cached) {
+                    // The process doe not exist anymore, but its
+                    // mutex is still cached - wait a bit before
+                    // reading the metadata to allow the non-blocking
+                    // to lock its mutex and update the metadata
+                    pcp_util::this_thread::sleep_for(
+                        pcp_util::chrono::milliseconds(METADATA_RACE_MS));
+                }
+            }
+        } catch (const std::invalid_argument& e) {
+            LOG_ERROR("Invalid value '%1%' stored in PID file '%2%'",
+                      pid_txt, pid_file);
+        }
+    }
+
     ActionMetadata metadata {};
+    auto metadata_file = (results_dir_path / "metadata").string();
 
     try {
-        metadata = ActionMetadata(metadata_file);
+        // Acquire the result mutex lock, if the transaction is cached
+        ResultsMutex::Mutex_Ptr mtx_ptr;
+        {
+            ResultsMutex::LockGuard a_l {
+                ResultsMutex::Instance().access_mtx };  // RAII
+            if (ResultsMutex::Instance().exists(t_id)) {
+                mtx_ptr = ResultsMutex::Instance().get(t_id);
+            }
+        }
+        if (mtx_ptr != nullptr) {
+            LOG_TRACE("Non-blocking transaction %1% mutex is cached; "
+                      "locking it", t_id);
+            ResultsMutex::LockGuard r_l { *mtx_ptr };  // RAII
+            metadata = ActionMetadata(metadata_file);
+        } else {
+            LOG_TRACE("Non-blocking transaction %1% mutex is not cached", t_id);
```

Comment:
I'm not sure I understand your question, because the non-blocking task that executes external modules (say, pxp module puppet) and uses ResultsMutex is https://github.com/puppetlabs/pxp-agent/blob/master/lib/src/request_processor.cc#L106; it's a different thread within the pxp-agent process - it doesn't have a different PID.

The non-blocking task is the one the writes the PID of the external module process to file, via leatherman's execution. But that PID file is stored in the spool directory named after the PXP transaction ID; so, if we assume that such UUID is unique, the writing operation is always safe, regardless of the PID value. But I guess this not is what you're concerned of...

As for accessing the metadata file, if the mutex named after the transaction ID is not cached in the ResultsMutex singleton, it means that the metadata will not be updated. But, regardless of that, the non-blocking action writes the metadata file with atomic_write_to_file(), so reading should always be safe. What we enforce with this commit is that the transaction status handler will wait for the non-blocking until the metadata is updated, when necessary - the transaction status handler only reads the metadata; concurrent transaction status handlers can operate safely.

Also, note that in ResultsMutex we have double locking with the singleton's access_mutex (that must be locked to access the cache) and the mutexes of the transaction currently being processed (that must be locked, when existent, to access the metadata file).

Hope this clarifies a bit.